### PR TITLE
Refactor DefaultLatexParser, LatexParserResult classes

### LINKED ISSUE #10585: Add FileMonitor for LaTeX citations
The [LaTeX citations](https://docs.jabref.org/advanced/entryeditor/latex-citations) are NOT backed by a file update monitor.

a

Rewrite the code to use our `org.jabref.gui.util.DefaultFileUpdateMonitor` (or a new file update monitor) with proper shutdown at the end.

(Note: This is the technically sound fix for https://github.com/JabRef/jabref/issues/10584)

---

**Update**

A first PR was submitted at https://github.com/JabRef/jabref/pull/10937. It turned out, that a **new** FileMonitor is needed. It should watch a directory.

- New .tex file: Parse the tex file, remove all non-existant citaton keys and add the new ones. Add the tex file to file listeners
- Deleted .tex file: Remove the .tex file from the file listeners

The file listeners:

- Changed.tex file: Parse the tex file, remove all non-existant citaton keys and add the new ones

---

- Class org.jabref.logic.texparser.DefaultLatexParser needs to be modified to know the found latex citation keys per latex file. In case an "udpate" of the informaiton is triggered (by the file update monitor!!), all existing keys should be dropped and readded. Thus, org.jabref.model.texparser.LatexParserResult also needs to adapted (see below)
- Refactor `org.jabref.gui.entryeditor.LatexCitationsTabViewModel`: Move away `org.jabref.gui.entryeditor.LatexCitationsTabViewModel#searchDirectory`, `org.jabref.gui.entryeditor.LatexCitationsTabViewModel#searchAndParse` as class `CitationFinder` in package `org.jabref.logic.texparser`. -- Also test cases need to be added
- Refactor `org.jabref.model.texparser.LatexParserResult` to allow partial updates. That means: If file X is changed, call new DefaultLatexParser().parse(X) and change the contents of `org.jabref.model.texparser.LatexParserResult#citations` accordingly. IMHO, a map from `Path` to `Set<Citations>`  (Multimap) should be kept. The value before the update and after the update be stored. A diff calculated. That diff should be applied to `citations`.  - The method `update` can be implemented in `LatexParserResult`. - The class `LatexParserResult` should be renamed in `CitationState`.
- A new file watcher, which works on directories and handles both file editing and creation of new files. On both events, `CitationState#update(path)` can be called. That method should simply add the file to the currently known files. -- Deletion has to be handled separately. event: delete, new method `CitationState#remove(path)` called

Reasons: Energy saving. Very little CPU usage for an update on one file. Although your current method works, it parses ALL `.tex` files again, and handles the content. Large LaTeX projects have dozens of `tex` files, which are very long. This costs time.

### PR DESCRIPTION
**Pull request by LoayGhreeb**
I'm working on implementing a new file monitor for LaTeX citations.

- I haven't added the file monitor itself yet, but I've made the necessary modifications to the parser and result classes.
- Updated `org.jabref.model.texparser.LatexParserResult` class to store the parsed results of a single .tex file.
- Introduced `org.jabref.model.texparser.LatexParserResults` class to collect all parsed .tex files (the class name may need to be changed).

Refs #10585

### Mandatory checks

- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Review by koppor — CHANGES_REQUESTED** [review]
Minor comments, otherwise, look good.

Need to try out, however ^^.

Proposal:

Get this in as fast as possible and implement the new file watcher in a follow-up PR. - I already changed `Closes` to `Refs` in the PR description

File: src/main/java/org/jabref/logic/texparser/DefaultLatexParser.java [hunk 1]
```diff
@@ -49,75 +49,78 @@ public class DefaultLatexParser implements LatexParser {
...
     public LatexParserResult parse(String citeString) {
-        matchCitation(Path.of(""), 1, citeString);
+        Path path = Path.of("");
+        LatexParserResult latexParserResult = new LatexParserResult(path);
+        matchCitation(path, 1, citeString, latexParserResult);
         return latexParserResult;
     }
 
     @Override
     public LatexParserResult parse(Path latexFile) {
```
[the remark this answers is not in the document]
Comment (LoayGhreeb) [reply src/main/java/org/jabref/logic/texparser/DefaultLatexParser.java:67]: This method will never return null. However, the `parse(Path latexFile)` method may return null if the path is not found. I have updated it to return Optional instead.

File: src/main/java/org/jabref/model/texparser/LatexParserResults.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+public class LatexParserResults {
+    private final Map<Path, LatexParserResult> parsedTexFiles;
+
+    public LatexParserResults() {
+        this.parsedTexFiles = new HashMap<>();
+    }
+
+    /**
+     * For testing purposes
+     */
```
[the remark this answers is not in the document]
Comment (LoayGhreeb) [reply src/main/java/org/jabref/model/texparser/LatexParserResults.java:24]: Yes, I used it in the test package, will add the annotation and keep the public, but the [documentation](https://guava.dev/releases/33.1.0-jre/api/docs/com/google/common/annotations/VisibleForTesting.html) says:

>Do not use this interface for public or protected declarations: it is a fig leaf for bad design.

**Review by koppor — CHANGES_REQUESTED** [review]
"Refresh" button does not update the line numbers of the found citatoin. I loaded the below example, then edited `demo.tex` by adding newlines above the `\cite` line. Pressed refresh. Nothing changed.

Tried it out with following files:

`demo.tex`

```latex
\documentclass{scrartcl}
\begin{document}
Demo \cite{Steinmacher2015}.
\bibliographystyle{unsrt}
\bibliography{demo}
\end{document}
```

`demo.bib`:

```bibtex
@InProceedings{Simon2019,
  author    = {Martin K. Simon and Linus W. Dietz and Tobias Diez and Oliver Kopp},
  booktitle = {Proceedings of the 11th Central European Workshop on Services and their Composition, Bayreuth, Germany, February 14-15, 2019},
  title     = {Analyzing the Importance of JabRef Features from the User Perspective},
  year      = {2019},
  editor    = {Stefan Kolb and Christian Sturm},
  pages     = {47--54},
  publisher = {CEUR-WS.org},
  series    = {{CEUR} Workshop Proceedings},
  volume    = {2339},
  bibsource = {dblp computer science bibliography, https://dblp.org},
  biburl    = {https://dblp.org/rec/conf/zeus/SimonDDK19.bib},
  groups    = {jabref},
  url       = {https://ceur-ws.org/Vol-2339/paper10.pdf},
}

@Article{Steinmacher2015,
  author    = {Igor Steinmacher and Marco Aurelio Graciotto Silva and Marco Aurelio Gerosa and David F. Redmiles},
  journal   = {Information and Software Technology},
  title     = {A systematic literature review on the barriers faced by newcomers to open source software projects},
  year      = {2015},
  issn      = {15230864},
  month     = {mar},
  pages     = {67--85},
  volume    = {59},
  doi       = {10.1016/j.infsof.2014.11.001},
  file      = {:IST2015.pdf:PDF},
  groups    = {software engineering},
  keywords  = {Open Source Software, Software Engineering, Newcomers, Beginners, Novices, Joining, Contribution, Barriers to Entry, Onboarding, Open Collaboration, Socialization, Systematic Literature Review},
  publisher = {Elsevier {BV}},
}

@Article{,
  title = {Demo},
  file  = {:- Demo.pdf:PDF},
}

@Comment{jabref-meta: databaseType:bibtex;}

@Comment{jabref-meta: fileDirectory:.;}

@Comment{jabref-meta: grouping:
0 AllEntriesGroup:;
1 StaticGroup:software engineering\;0\;1\;0x008000ff\;\;\;;
1 StaticGroup:jabref\;0\;1\;0xff00ffff\;\;\;;
1 TexGroup:Zitiert\;0\;C:/Users/olive/OneDrive/Projects/JabRef DANTE 2024/demo/demo.aux\;1\;\;\;\;;
}
```

**Comment by koppor** [comment]
Double check: Using `main` and the above files, the `Refresh` button works.

**Comment by LoayGhreeb** [comment]
Good catch, I missed updating the `citationFinder` in the button action. 
By the way, what is the better way to deal with this: updating the path in the `citationFinder` to be `null` or just creating a new `CitationFinder`? I also have to do this in the `checkAndUpdateDirectory()`. (create a new object or update).

**Comment by koppor** [comment]
> Good catch, I missed updating the `citationFinder` in the button action. By the way, what is the better way to deal with this: updating the `latexParserResults` in the `citationFinder` to be `null` or just creating a new `CitationFinder`? I also have to do this in the `checkAndUpdateDirectory()`. (create a new object or update the path and set the `latexParserResults` to null).

If have no real clue about the inner workings... I remember that using `null` was kind of lazy loading. However, maybe, recreation the parser early is also OK (because it does not take much time to do so).

**Review by koppor — APPROVED** [review]
Nice refactorings!

Code works fine. Also for non-existing directories:

image

**Pull request merged**